URW-180 PayPal API

Kenneth-W-Chen opened the pull request:
# Summary
* Uses PayPal REST API since PayPal deprecated their ButtonManager and NVP/SOAP APIs
  * `PayPalCustomApi` (`paypal.php`) serves as an intermediary for the API. Currently implements getting an OAuth token, creating orders, getting order details, and capturing payment.
  * Created some objects to represent API request body information (`paypal/api-json-objects.php`)
* Orders are created when the user presses the PayPal button. 
  * Users are sent to `/paypal/get-order` which sets the `Location` header to the PayPal checkout link.
* IPN replaced with webhook event handling at `/paypal/event-notify`.
  * IPN handlers have been moved to `webhook/handlers`. The functionality remains mostly the same.
## Missing functionalities (won't be added in this ticket)
* Integration with API endpoints that are not `orders` and `webhooks`
  * Missing from `orders`: `Update Orders`, `Confirm the Order`, `Authorize payment for order`, `Add tracking information for an Order`, `Update or cancel tracking information for a PayPal Order`.
  * Missing from `webhooks`: everything except `Verify webhook signature`.
* Payment methods besides PayPal are not implemented. Apple/Google Pay won't be implemented in this PR.
* Removed order info output on the `/merch/buy/complete` page because
  * Payment details aren't available at this point and order ID isn't usable by the customer
  * Retrieving product info would require a call to the PayPal and Printful APIs

## Other changes
* `config.php` 
  * Defined constants `PAYPAL_SANDBOX_API_CLIENT_ID`, `PAYPAL_SANDBOX_API_SECRET_KEY_1`, `PAYPAL_SANDBOX_API_ACCOUNT`, `PAYPAL_SANDBOX_API_PASSWORD`, `PAYPAL_SANDBOX_API_SIGNATURE`, and `PAYPAL_SANDBOX_WEBHOOK_ID`
* Removed `pointer-events: none` from `.paypal-button-overlay` for the PayPal button since the button is now an actual button
* Changed receipt email flow to only send one receipt instead of multiple per transaction.
  * Moved it to a separate function in `/template/functions/paypal.php`.
  * The email is sent as an attachment to the web-dev-logs channel if it's in the dev environment.
* Dues T-shirt discount is now actually shown on the PayPal receipt as a separate field. The amount paid is still the same.
* Removed an error thrown in the Printful handler from when the revenue is less than the total cost of the product. 
  * This ends up being the case most of the time because of shipping+tax exceeding Printful's suggested retail price
  * Swapped it to a discord message that also lists the shipping and tax

## Testing
- [x] Test `PayPalCustomApi` functionality
- [x] Test `PayPalWebhookEvent` functionality
- [x] Test button creation on frontend 
- [x] Test webhook event filtering works
- [x] Test that order creation works through the button
- [x] Test the dues webhook (formerly IPN) handler
- [x] Test the Printful webhook (formerly IPN) handler
  - [x] Test that Printful items are added to db properly (`/merch/product` and function `insert_paypal_item()`)
  - [x] Test that Printful orders are properly made
- [x] Check that the email HTML works
  - [x] Check Dues-only receipt
  - [x] Check Printful-only receipt
  - [x] Check Dues and Printful receipt

# Order process (flowchart updated 10/11/2024)
![paypal-api-flow](https://github.com/user-attachments/assets/c9f93f8d-049e-4d58-ab45-804f340bc111)

## API minor notes
* If you create an order that needs shipping info but don't provide it, the response from create_order will return a checkout link where the payer will provide their shipping info and payment info. Payment source info in `capture_payment` does not need to be filled out in this case (although you still need to pick a payment source [e.g., PayPal or Apple Pay]).
* Tax isn't a necessary field for the API.
## Receipt notes
* Technically the buyer receives a receipt from PayPal too with the items they bought. Except for Printful information (IDs and shipping), everything else is likely duplicated
* "Order Name" will be a comma-delimited list of the order items. E.g., "Dues, Dues Shirt" will be the order name
# Usage
### Create orders
Create a `PayPalCustomApi` object and call `create_order()` with the appropriate arguments. The response is JSON-encoded string with the order ID, status, and links to checkout and do other things.
Example response:
```json
{
  "id": "123ABC456DEF",
  "status": "PAYER_ACTION_REQUIRED",
  "payment_source": {
    "paypal": {}
  },
  "links": [
    {
      "href": "https://api.sandbox.paypal.com/v2/checkout/orders/123ABC456DEF",
      "rel": "self",
      "method": "GET"
    },
    {
      "href": "https://www.sandbox.paypal.com/checkoutnow?token=123ABC456DEF",
      "rel": "payer-action",
      "method": "GET"
    }
  ]
}
```
### Capture payments
Call this function on an order ID after the payer has approved payment.
```php
require_once('../api/paypal/paypal.php');

$p = new PayPalCustomApi();
$p->capture_payment($_GET['token'],true);
```
You should check the `status` field returned by the capture request. Specifically, you want `COMPLETED`. See [the docs](https://developer.paypal.com/docs/api/orders/v2/#orders_capture!c=201&path=status&t=response) for more details.
### Get order info
Call this on an order ID to get information on the order from PayPal.
```php
require_once('path/to/dir/api/paypal/paypal.php');
$p = new PayPalCustomApi();
$r = $p->get_order_info('orderid');
```
The result will be a JSON-encoded string. It's a bit long, so I'm not including a sample here.

### Creating a payment button
```php
require_once('path/to/dir/template/functions/paypal.php');
$items = [['type':'dues', 'full_year':true, 't-shirt':false], ['type':'printful','ext_id':'externalid' , 'variant_id':'variantid'], ['type':'donation', 'amount':'99.99'], ...];
get_payment_button('text to display', $items, 'return/page/uri', 'cancel/page/uri');
```
* `text to display` is the text displayed on the button. All letters will be capitalized.
* `$items` is an array of the items to send. The format for all implemented types is shown.
  *  The array should be a 2-D array with each 1-D array representing a single item.
  * To change a value with JavaScript, access the input's value with `input#paypal-items<index><key>` e.g., `input#paypal-items0full_year`
* `return/page/uri` is the page to return the user to upon completing (approving) the order. This is relative to the website's server name (e.g., www.untrobotics.com)
* `cancel/page/uri` is the page to return the user to upon cancelling the order. It follows the same rules as `return/page/uri` 

# To-do
* Cache the bearer token (requires #158 to be merged)

Kenneth-W-Chen committed:
Functionality to get PayPal access token

Kenneth-W-Chen committed:
Created some classes corresponding to JSON objs stuff from the API (a lot of this may be unnecessary)

Kenneth-W-Chen committed:
Added create_order function. Added some constructors for the JSON classes.
Minor changes to PHPDocs

Kenneth-W-Chen committed:
Added payment capturing, adjusted some docs, created a class for confirm_order (unused)
Refactoring, created PayPalApplicationContext class, adjusted some error logging

Kenneth-W-Chen commented:
7bf6ff4 adds a helper function that's in #158. Once that's merged, I'll remove the one at the bottom of `api/paypal/paypal.php` and call the other one. I'll also run the requests through the caching system, although the only thing we'll cache is probably info from `get_order`

Kenneth-W-Chen committed:
Added different shipping options to create_order
Catch error from send_request in create_order

Kenneth-W-Chen committed:
Create webhook event verify via API
Fixed issue with create_order that threw an unexpected error

Kenneth-W-Chen commented:
There's a way to [verify webhook events without PayPal's API](https://github.com/untrobotics/website/pull/162/commits/ed0bbdec9ddb594dae2dd719a48c393318c1a18a#diff-f7b4a23d976318b87279ab29da50efd130fa58408f8d6e8f226e90ffe95c6400R40-R65), but it does require PayPal's certificate. Is there a folder we store certs in already or should I store it in a new folder?

Kenneth-W-Chen committed:
Create basic webhook handler; create basic table structure to store PayPal transaction items

Kenneth-W-Chen commented on sql/paypal_transactions.sql:
```diff
@@ -0,0 +1,17 @@
+    PRIMARY KEY (`id`)
+) ENGINE=InnoDB AUTO_INCREMENT=47 DEFAULT CHARSET=latin1;
+
+/*CREATE TABLE `paypal_orders` (
+    `id` int(11) NOT NULL AUTO_INCREMENT,
+    `paypal_id` varchar(255) NOT NULL,                                  -- The order ID PayPal generates
+)*/
```
I can't think of a good use case for this since PayPal always (from what I have seen) returns the full order info if an order ID is in the response.
I can make it with 10 foreign keys from `paypal_transactions`, (PayPal restricts the number of items per order to 10). But, again, use case?

Kenneth-W-Chen commented on sql/paypal_transactions.sql:
```diff
@@ -0,0 +1,17 @@
+CREATE TABLE `paypal_transactions` (
+    `id` int(11) NOT NULL AUTO_INCREMENT,
+    `custom_id`             varchar(255) NOT NULL,  -- This is the ID that PayPal receives for transactions; we generate this
```
https://developer.paypal.com/docs/api/orders/v2/#orders_get!c=200&path=purchase_units/custom_id&t=response

Kenneth-W-Chen commented on sql/paypal_transactions.sql:
```diff
@@ -0,0 +1,17 @@
+CREATE TABLE `paypal_transactions` (
```
I'm not sure if "transactions" is the proper term. Maybe "items" or "order_items" since `custom_id` is for each item?

Kenneth-W-Chen commented:
I stopped having fun about 30 minutes ago

Kenneth-W-Chen committed:
Make table and funcs to return payment order info to button

Kenneth-W-Chen committed:
Backend implementation to create orders when button is pressed

Kenneth-W-Chen committed:
Added discounts to items since discounted t-shirt

Kenneth-W-Chen committed:
Fixed pass by reference issue, removed catch clause from create_order
Changed some PHPDocs

Kenneth-W-Chen committed:
Button now works

Kenneth-W-Chen committed:
Add FK to user.id for transactions

Kenneth-W-Chen committed:
Improved documentation. Reformatted files

Kenneth-W-Chen committed:
Added Dues button. Doesn't work for t-shirts yet

Kenneth-W-Chen commented:
Me having fun now

Kenneth-W-Chen committed:
Add drop statements

Kenneth-W-Chen committed:
Added webhook event handling

Kenneth-W-Chen committed:
Define event to remove expired orders every 3 days

Kenneth-W-Chen committed:
Fixed button not showing in the Spring semester

Kenneth-W-Chen committed:
Updated Dues handler.
Receipt email is only sent once now.
Removed $fields parameter from get_order_info since you can only filter for "payment_source" (which only returns the order ID).
Added variant_name to paypal_items table. Added variant stuff to the order creation process
Added function to tell if a webhook event is sandbox.
Stopped webhook handler from trying to capture payment we already marked as captured
Made a new payment button function for when the page has static info on the order (i.e., Printful products and dues page during the Spring)
Moved the IPN handlers to webhook/handlers/

Kenneth-W-Chen commented:
I should've made the commit message for f58829b say "Fixed some stuff"

Kenneth-W-Chen committed:
Fixed order button not working with multiple items/variant names

Kenneth-W-Chen committed:
Dues payments are now working

Kenneth-W-Chen committed:
Fixed button not being centered on dues page

Kenneth-W-Chen committed:
Fixed receipt email

Kenneth-W-Chen committed:
Improved comments and PHPDocs

Kenneth-W-Chen committed:
Add info to db for printful products

Kenneth-W-Chen committed:
Fix undefined variant name

Kenneth-W-Chen committed:
Create PayPal order with Printful products

Kenneth-W-Chen committed:
Update Printful handler

Kenneth-W-Chen committed:
Add Printful handler to webhook handler

Kenneth-W-Chen committed:
Remove references to IPN

Kenneth-W-Chen committed:
Printful handler works now

Kenneth-W-Chen committed:
Fix errors in merch buy complete page

Kenneth-W-Chen committed:
Fix dues shirt external IDs

Kenneth-W-Chen committed:
Change brand name to say UNT Robotics by default

Kenneth-W-Chen committed:
Send email in discord if in dev environment
Typos and remove unnecessary messages

Kenneth-W-Chen committed:
Change wording for dues+shirt payment

sebastian-king commented on sponsorships/index.php:
```diff
@@ -21,7 +22,7 @@
         </div>
         <div class="cell-lg-12 offset-lg-top-50">
-            <?php echo $payment_button->get_button()->button; ?>
+<!--            todo: add payment button-->
```
This is still a TODO

Kenneth-W-Chen committed:
Add UPC constructor

Kenneth-W-Chen committed:
Rewrite payment button frontend

Kenneth-W-Chen committed:
Rewrite payment button backend
Minor bug fixes

Kenneth-W-Chen committed:
Add donation button. Change get_payment_button params

Kenneth-W-Chen committed:
Verify webhook events internally before externally

Kenneth-W-Chen committed:
Adjust stuff for new data location

Kenneth-W-Chen committed:
Fix Printful handler stuff. Remove some commented-out code

Kenneth-W-Chen committed:
Fix Printful order handling. Remove sql tables for paypal stuff

Kenneth-W-Chen committed:
Adjust PayPalWebhookEvent to pick an API URL based on the event
Email receipt Discord message now puts the email body in a separate html file
Minor comment changes/removals

Kenneth-W-Chen committed:
Merge develop into URW-180-Paypal-API

Kenneth-W-Chen committed:
Pass PayPal API requests through the caching system

sebastian-king commented:
Closing as part of reconciling prod drift. The work in these PRs was based on `develop`, which has since fallen ~282 commits behind the actual prod state (now on `chore/reconcile-prod-drift`). Prod already implements this functionality — generally in a newer/reworked form — so this PR is superseded. The branch is preserved, so nothing is lost and anything still useful can be cherry-picked. We're auditing each of these for any unique change worth incorporating before the reconcile branch merges into develop.
